src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultVisualStudioOpenDocumentManager.cs

```diff
@@ -0,0 +1,167 @@
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            if (workspace == null)
+            {
+                throw new ArgumentNullException(nameof(workspace));
+            }
+
+            if (editorAdaptersFactoryService == null)
+            {
+                throw new ArgumentNullException(nameof(editorAdaptersFactoryService));
+            }
+
+            _fileChangeService = serviceProvider.GetService(typeof(SVsFileChangeEx)) as IVsFileChangeEx;
+            _runningDocumentTable = serviceProvider.GetService(typeof(SVsRunningDocumentTable)) as IVsRunningDocumentTable;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+
+            _documents = new List<VisualStudioDocumentTracker>();
+            _viewImportChangeTrackerCache = new Dictionary<string, ViewImportChangeTracker>(StringComparer.OrdinalIgnoreCase);
+
+            _foregroundDispatcher = workspace.Services.GetRequiredService<ForegroundDispatcher>();
+            _errorReporter = workspace.Services.GetRequiredService<ErrorReporter>();
+        }
+
+        public override IReadOnlyList<VisualStudioDocumentTracker> OpenDocuments => _documents;
+
+        public override void AddDocument(VisualStudioDocumentTracker tracker)
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            if (!_documents.Contains(tracker))
+            {
+                _documents.Add(tracker);
+
+                if (!tracker.TextBuffer.Properties.TryGetProperty<VisualStudioRazorParser>(typeof(VisualStudioRazorParser), out var parser))
+                {
+                    // The document tracker doesn't have a corresponding Razor parser. This should never be the case.
+                    return;
+                }
+
+                EnsureUndisposedTrackers();
+
+                var imports = tracker.Imports;
+                foreach (var import in imports)
+                {
+                    var importFilePath = import.FilePath;
+                    if (importFilePath == null)
+                    {
+                        // This is probably an in-memory view import. We can't track it.
+                        continue;
+                    }
+
+                    ViewImportChangeTracker viewImportChangeTracker;
+                    if (!_viewImportChangeTrackerCache.ContainsKey(importFilePath))
+                    {
+                        // First time seeing this view import. Create a change tracker for it.
+                        viewImportChangeTracker = new ViewImportChangeTracker(
+                            _fileChangeService,
+                            _runningDocumentTable,
+                            _editorAdaptersFactoryService,
+                            _foregroundDispatcher,
+                            _errorReporter,
+                            importFilePath);
```

Comment:
Consider making these 'dumb' - like just have all the data. Then you don't need to pass around all of these services.